webapp/src/Entity/Problem.php

```diff
@@ -287,26 +291,84 @@ public function getSpecialCompareArgs(): ?string
         return $this->special_compare_args;
     }
 
-    public function setCombinedRunCompare(bool $combinedRunCompare): Problem
+    public function setTypesAsString(array $types): Problem
     {
-        $this->combined_run_compare = $combinedRunCompare;
+        $stringToTypes = array_flip($this->typesToString);
+        $typeConstants = [];
+        foreach ($types as $type) {
+            if (!isset($stringToTypes[$type])) {
+                throw new Exception("Unknown problem type: '$type', must be one of " . implode(', ', array_keys($stringToTypes)));
+            }
+            $typeConstants[$type] = $stringToTypes[$type];
+        }
+        $this->setTypes($typeConstants);
+
         return $this;
     }
 
-    public function getCombinedRunCompare(): bool
+    public function getTypesAsString(): string
+    {
+        $typeConstants = $this->getTypes();
+        $typeStrings = [];
+        foreach ($typeConstants as $type) {
+            if (!isset($this->typesToString[$type])) {
+                throw new Exception("Unknown problem type: '$type'");
+            }
+            $typeStrings[] = $this->typesToString[$type];
+        }
+        return implode(', ', $typeStrings);
+    }
+
+    public function getTypes(): array
     {
-        return $this->combined_run_compare;
+        $ret = [];
+        foreach (array_keys($this->typesToString) as $type) {
+            if ($this->types & $type) {
+                $ret[] = $type;
+            }
+        }
+        return $ret;
     }
 
-    public function setMultipassProblem(bool $isMultipassProblem): Problem
+    public function setTypes(array $types): Problem
     {
-        $this->isMultipassProblem = $isMultipassProblem;
+        $types = array_unique($types);
+        $this->types = 0;
+        foreach ($types as $type) {
+            $this->types |= $type;
+        }
+        if ($this->types & self::TYPE_PASS_FAIL && $this->types & self::TYPE_SCORING) {
```

Comment:
Wouldn't this be better?
```suggestion
        if (($this->types & self::TYPE_PASS_FAIL) xor ($this->types & self::TYPE_SCORING)) {
```

Reply:
I must be missing something. If I am not mistaken, my code showed the warning if both pass-fail and scoring bits are set while your code shows the warning in case exactly one is set?

Reply:
Indeed. But isn't what I propose the correct constraint to enforce? A problem should always be either pass-fail or scoring. Never neither or both? But maybe I'm missing context of how this is handled in our code.

Reply:
Understood, I changed your code suggestion by negating it, and adjusting the error message.